### PR TITLE
[1.0]client/db: enable archive pruning

### PR DESCRIPTION
**Pull request by buck54321**
This is a refactor of #2975 meant as a stop-gap for the release branch. I will close #2975 in favor of a more comprehensive archiving solution to be worked out later.

### CODE REVIEW

File: client/db/bolt/db.go [hunk 1]
```diff
@@ -408,6 +414,125 @@ func (db *BoltDB) SetPrimaryCredentials(creds *dexdb.PrimaryCredentials) error {
...
+			oidsWithActiveMatches[oid] = struct{}{}
+			return nil
+		}); err != nil {
+			return fmt.Errorf("error building active match order ID index: %w", err)
+		}
+
+		nOrds := uint64(archivedOB.Stats().BucketN - 1 /* BucketN includes top bucket */)
+		if nOrds <= prunedSize {
+			return nil
+		}
```
Comment (JoeGruffins) [inline client/db/bolt/db.go:447]: Could check this first and avoid getting the `oidsWithActiveMatches` sometimes

File: client/app/config.go [hunk 2]
```diff
@@ -106,6 +107,8 @@ type CoreConfig struct {
 	UnlockCoinsOnLogin bool `long:"release-wallet-coins" description:"On login or wallet creation, instruct the wallet to release any coins that it may have locked."`
 
 	ExtensionModeFile string `long:"extension-mode-file" description:"path to a file that specifies options for running core as an extension."`
+
+	PruneArchive uint64 `long:"prunearchive" description:"prune that order archive to the specified number of most recent orders"`
```
Comment (JoeGruffins) [inline client/app/config.go:111]: Could add that zero disables pruning.

**Review by JoeGruffins — APPROVED** [review]

File: client/db/bolt/db.go [hunk 3]
```diff
@@ -408,6 +414,125 @@ func (db *BoltDB) SetPrimaryCredentials(creds *dexdb.PrimaryCredentials) error {
...
+			}
+			stamp := intCoder.Uint64(stampB)
+			if len(deletes) < toClear {
+				deletes = append(deletes, &orderStamp{
+					stamp: stamp,
+					oid:   oidB,
+				})
+				sortDeletes()
+				return nil
+			}
```
Comment (martonp) [inline client/db/bolt/db.go:484]: Could avoid sorting until  `len(deletes) == toClear`:

```suggestion
			if len(deletes) < toClear {
				deletes = append(deletes, &orderStamp{
					stamp: stamp,
					oid:   oidB,
				})
				return nil
			}
			if !sorted {
			    sortDeletes()
			    sorted = true
			}
```

**Review by martonp — APPROVED** [review]

**Review by dev-warrior777 — APPROVED** [review]
Looks good after review changes

**Pull request merged**